src/bookshelf/src/bksw_io.erl

```diff
@@ -117,12 +121,43 @@ entry_delete(FullPath) ->
     case file:delete(FullPath) of
         ok ->
             ?LOG_DEBUG("deleted bucket entry: ~p", [FullPath]),
-            true;
+            entry_dir_delete(FullPath);
         Error ->
             error_logger:error_msg("Error deleting bucket entry ~p: ~p~n", [FullPath, Error]),
             false
     end.
 
+%%
+%% We potentially create a directory for every file we store in bookshelf. This can grow pretty
+%% large (4 levels deep, with 256 entries per level, so 2^32 + 2^24 + 2^16 + 2^8 > 4 billion). If we
+%% never delete them we will blow out the inode table, and consume considerable disk space. For
+%% example ext4 will most likely allocate a full block (default 4KiB) for each directory we create
+%% due to the long file names for the files we store
+%% (https://ext4.wiki.kernel.org/index.php/Ext4_Disk_Layout)
+%%
+%% The depth parameter controls how aggressive we are. A good value would be 3, leaving the upper
+%% level intact. If we were willing to sacrifice an extra 64k directories, '2' could be chosen
+%% instead for some minor performance gain.
+entry_dir_delete(FullPath) -> entry_dir_delete(FullPath, ?DIRECTORY_CLEANUP_DEPTH).
+
+entry_dir_delete(FullPath, 0) ->
+    true;
+entry_dir_delete(FullPath, Depth) ->
+    DirName = file:dirname(FullPath),
+    %% Speculatively delete the directory, expecting it to fail if it's not empty.
+    case file:del_dir(DirName) of
+        true ->
+            %% Succeeded, let's go try the next directory down.
+            ?LOG_DEBUG("deleted directory entry: ~p", [DirName]),
+            entry_dir_delete(DirName, Depth-1);
+        {error, eexist} ->
+            %% Not empty, which can reasonably happen, so just quit
+            true;
+        {error, Error} ->
```

Comment:
According to the file:de_dir docs this might return:

```
                enoent:
                  The directory does not exist.
```

which feels like it is possible in the normal case if we were to get concurrent delete requests.  Perhaps we should return true there as well like we do for eexists?  In the current code it doesn't seem to matter too much since we don't look at the return code fromt his function.